src/cp/sliprules.cxx

```diff
@@ -0,0 +1,189 @@
+#include "sliprules.h"
+
+namespace neml {
+
+double SlipRule::sum_slip(const Symmetric & stress, const Orientation & Q, 
+                          const History & history, Lattice & L, 
+                          double T) const
+{
+  double dg = 0.0;
+  for (size_t g = 0; g < L.ngroup(); g++) {
+    for (size_t i = 0; i < L.nslip(g); i++) {
+      dg += fabs(slip(g, i, stress, Q, history, L, T));
```

Comment:
I think in this case there is no asymmetry in slip (+gamma_dot = -gamma_dot ) and both contribute to shear, but in some models there can be an asymmetry, meaning if resolved shear stress (rss) is negative there would be no slip. I am not sure whether adding/commenting on such is a good idea here, but we can write something like dg = +dg + (-dg) , where in this case -dg = +dg.

Reply:
The only purpose of this routine is to give you the sum of the absolute slip rates for hardening models that want that info.  The abs is correct for this use.

Or is your question about bi-directional slip?  I have these setup so that the slip() function will return a positive or negative value depending on the sign of the resolved shear.  So again, this is correct here.